java/src/main/native/src/CuFileJni.cpp

```diff
@@ -0,0 +1,238 @@
+#include <unistd.h>
+#include <memory>
+
+#include <stdexcept>
+
+#include <cstring>
+
+char const* GetCuErrorString(CUresult cu_result)
+{
+  char const* description;
+  if (cuGetErrorName(cu_result, &description) != CUDA_SUCCESS) description = "unknown cuda error";
+  return description;
+}
+
+std::string cuFileGetErrorString(int error_code)
+{
+  return IS_CUFILE_ERR(error_code) ? std::string(CUFILE_ERRSTR(error_code))
+                                   : std::string(std::strerror(error_code));
+}
+
+std::string cuFileGetErrorString(CUfileError_t status)
+{
+  std::string error = cuFileGetErrorString(status.err);
+  if (IS_CUDA_ERR(status)) { error.append(".").append(GetCuErrorString(status.cu_err)); }
+  return error;
+}
+
+/** @brief RAII wrapper for a file descriptor and the corresponding cuFile handle. */
+class cufile_file {
+ public:
+  /**
+   * @brief Construct a file wrapper.
+   *
+   * Should not be called directly; use the following factory methods instead.
+   *
+   * @param file_descriptor A valid file descriptor.
+   */
+  explicit cufile_file(int file_descriptor) : file_descriptor_{file_descriptor}
+  {
+    CUfileDescr_t cufile_descriptor{CU_FILE_HANDLE_TYPE_OPAQUE_FD, file_descriptor_};
+    auto const status = cuFileHandleRegister(&cufile_handle_, &cufile_descriptor);
+    if (status.err != CU_FILE_SUCCESS) {
+      close(file_descriptor_);
+      throw std::logic_error("Failed to register cuFile handle: " + cuFileGetErrorString(status));
+    }
+  }
+
+  /**
+   * @brief Factory method to create a file wrapper for reading.
+   *
+   * @param path Absolute path of the file to read from. This file must exist.
+   * @return std::unique_ptr<cufile_file> for reading.
+   */
+  static auto make_reader(char const* path)
+  {
+    auto const file_descriptor = open(path, O_RDONLY | O_DIRECT);
+    if (file_descriptor < 0) {
+      throw std::logic_error("Failed to open file to read: " + cuFileGetErrorString(errno));
+    }
+    return std::make_unique<cufile_file>(file_descriptor);
+  }
+
+  /**
+   * @brief Factory method to create a file wrapper for writing.
+   *
+   * @param path Absolute path of the file to write to. This creates the file if it does not already
+   * exist..
+   * @return std::unique_ptr<cufile_file> for writing.
+   */
+  static auto make_writer(char const* path)
+  {
+    auto const file_descriptor = open(path, O_CREAT | O_WRONLY | O_DIRECT, S_IRUSR | S_IWUSR);
+    if (file_descriptor < 0) {
+      throw std::logic_error("Failed to open file to write: " + cuFileGetErrorString(errno));
+    }
+    return std::make_unique<cufile_file>(file_descriptor);
+  }
+
+  // Disable copy (and move) semantics.
+  cufile_file(cufile_file const&)            = delete;
+  cufile_file& operator=(cufile_file const&) = delete;
+
+  /** @brief Destroy the file wrapper by de-registering the cuFile handle and closing the file. */
+  ~cufile_file()
+  {
+    cuFileHandleDeregister(cufile_handle_);
+    close(file_descriptor_);
```

Comment:
Maybe print a warning if `close` fails?

Reply:
I would prefer to avoid printing to std::cerr from here. I think it would be good in a future PR to add some improved logging in general though and this could be one spot for that